src/main/java/ee/tuleva/onboarding/payment/PaymentProviderCallbackService.java

```diff
@@ -35,20 +35,29 @@ public void processToken(String serializedToken) {
 
     PaymentReference internalReference = getInternalReference(serializedInternalReference);
 
-    if (paymentRepository.findByInternalReference(internalReference.getUuid()).isEmpty()) {
+    if (paymentWithReferenceExists(internalReference) && isPaymentFinalized(token)) {
       User user = userService.findByPersonalCode(internalReference.getPersonalCode()).orElseThrow();
 
       Payment paymentToBeSaved =
           Payment.builder()
               .amount(amount)
               .internalReference(internalReference.getUuid())
               .user(user)
+              .status(PaymentStatus.PENDING)
               .build();
 
       paymentRepository.save(paymentToBeSaved);
     }
   }
 
+  private boolean paymentWithReferenceExists(PaymentReference internalReference) {
+    return paymentRepository.findByInternalReference(internalReference.getUuid()).isEmpty();
+  }
```

Comment:
`paymentWithReferenceExists == findByInternalReference isEmpty()`? Isn't it contradictory?

If it's empty then it _doesn't_ exist.

Reply:
fixed, thanks